pgoapi/pgoapi.py

```diff
@@ -686,6 +709,65 @@ def encounter_pokemon(self, pokemon_data, retry=False):  # take in a MapPokemon
             self.log.info("Could not start encounter for pokemon: %s", pokemon)
         return False
 
+    def incubate_eggs(self):
+        if not self.EGG_INCUBATION_ENABLED:
+            return
+        if self.player_stats.km_walked > 0:
+            for incubator in self.inventory.incubators_busy:
+                incubator_egg_distance = incubator['target_km_walked'] - incubator['start_km_walked']
+                incubator_distance_done = self.player_stats.km_walked - incubator['start_km_walked']
+                if incubator_distance_done > incubator_egg_distance:
+                    if not self.attempt_finish_incubation():
+                        break
+                else:
+                    self.log.info('Incubating %skm egg, %skm done', incubator_egg_distance, round(incubator_distance_done, 2))
+        for incubator in self.inventory.incubators_available:
+            if incubator['item_id'] == 901:  # unlimited use
+                pass
+            elif self.USE_DISPOSABLE_INCUBATORS and incubator['item_id'] == 902:  # limited use
+                pass
+            else:
+                continue
+            eggs_available = self.inventory.eggs_available
+            eggs_available = sorted(eggs_available, key=lambda egg: egg['creation_time_ms'],
+                                    reverse=False)  # oldest first
+            eggs_available = sorted(eggs_available, key=lambda egg: egg['egg_km_walked_target'],
+                                    reverse=self.INCUBATE_BIG_EGGS_FIRST)  # now sort as defined
+            if not self.attempt_start_incubation(eggs_available[0], incubator):
+                break
+
+    def attempt_start_incubation(self, egg, incubator):
+        self.log.info("Start incubating %skm egg", egg['egg_km_walked_target'])
+        incubate_res = self.use_item_egg_incubator(item_id=incubator['id'], pokemon_id=egg['id']).call()['responses']['USE_ITEM_EGG_INCUBATOR']
+        status = incubate_res.get('result', -1)
+        sleep(3)
+        if status == 1:
+            self.log.info("Incubation started with %skm egg !", egg['egg_km_walked_target'])
+            self.update_player_inventory()
+            return True
+        else:
+            self.log.debug("Could not start incubating %s", incubate_res)
+            self.log.info("Could not start incubating %s egg | Status %s", egg['egg_km_walked_target'], status)
+            self.update_player_inventory()
+            return False
+
+    def attempt_finish_incubation(self):
+        self.log.info("Checking for hatched eggs")
+        hatch_res = self.get_hatched_eggs().call()['responses']['GET_HATCHED_EGGS']
+        status = hatch_res.get('success', -1)
+        sleep(3)
+        if status == 1:
+            # still trying to find out how to get the values right, including the pokemon
```

Comment:
without understanding you did an awesome job! so the  
map will apply the lambda on each element of the list and will turn a list of ids in to a list of names then the join will join them with a , so we can print it out. if you have get_pokemon_by_id it would be just map(lambda x: get_pokemon_by_id(x), res.get('pokemon_id', []))